controllers/pipeline/pipeline_adapter.go

```diff
@@ -100,6 +100,15 @@ func (a *Adapter) EnsureSnapshotExists() (reconciler.OperationResult, error) {
 		a.logger.Info("Found existing Snapshot",
 			"snapshot.Name", existingSnapshot.Name,
 			"snapshot.Spec.Components", existingSnapshot.Spec.Components)
+		// Annotate the build pipelineRun with the existing Snapshot if it hasn't been already annotated
+		if _, found := a.pipelineRun.ObjectMeta.Annotations[tekton.SnapshotNameLabel]; !found {
+			a.pipelineRun, err = a.annotateBuildPipelineRunWithSnapshot(a.pipelineRun, existingSnapshot)
+			if err != nil {
+				a.logger.Error(err, "Failed to update the build pipelineRun with new annotations",
+					"pipelineRun.Name", a.pipelineRun.Name)
+				return reconciler.RequeueWithError(err)
```

Comment:
2 questions:

1. So now we give the controller **one more** chance to attach this annotation to the build PLR before we requeue the reconciliation loop with an error, right?
2. Consider the scenario when for a given PLR a snapshot was created but there was an issue during adding the annotation. So, when this PLR gets reconciled the next time, let's say that it is now not the most recently succeeded PLR so it will exit this function on line 87 above. This would result in some build PLRs not having this annotation, would that be a problem?

Reply:
I'd say that the chances of that are getting quite slim at that point - requeueing with error means that the controller will attempt another reconcile for the object within seconds (or less), so there would need to be a seriously quick succession of builds of the same component for that to happen.
In any case, if there is indeed a newer build for the component, then the Snapshot that we're trying to reconcile would get superseded by the new build anyways.

Reply:
> then the pipelineRun/Snapshot combination that we're trying to reconcile would get superseded by the new build anyways.

yup, and the previous build PLR would still not have the annotation on it, but that's not a problem I think.